### PR TITLE
WIP[KHWEBA-19] add tags

### PR DESCRIPTION
**Pull request by Vladyslav-Plakhuta**
screenshot

### CODE REVIEW

File: src/components/app/app.js [hunk 1]
```diff
@@ -4,8 +4,14 @@ import { DrawerComponent } from "../drawer/drawer";
 import { ToolbarComponent } from "../toolbar/toolbar";
 import { TransactionsComponent } from "../transactions/transactions";
 import { AddAccountComponent } from "../add-account-dialog/add-account-dialog";
-import { AddTagComponent } from "../add-tag-dialog/add-tag-dialog";
+import { AddTagComponent } from "../tags/add-tag-dialog/add-tag-dialog";
 import { AboutComponent } from "../about-dialog/about-dialog";
+import { PieChartComponent } from "../pie-chart/pie-chart";
+import { BarChartComponent } from "../bar-chart/bar-chart";
+import { TableTransactionsComponent } from "../table-transactions/table-transactions";
```
Comment (gift-a) [inline src/components/app/app.js]: charts and table are in transactions component

File: src/components/app/app.js [hunk 2]
```diff
@@ -4,8 +4,14 @@ import { DrawerComponent } from "../drawer/drawer";
...
+import { StorageService } from "../../services/storage";
+import { Tag } from "../../model/tag";
```
Comment (gift-a) [inline src/components/app/app.js]: we decided not to use models

File: src/components/app/app.js [hunk 3]
```diff
@@ -4,8 +4,14 @@ import { DrawerComponent } from "../drawer/drawer";
 import { ToolbarComponent } from "../toolbar/toolbar";
 import { TransactionsComponent } from "../transactions/transactions";
 import { AddAccountComponent } from "../add-account-dialog/add-account-dialog";
-import { AddTagComponent } from "../add-tag-dialog/add-tag-dialog";
+import { AddTagComponent } from "../tags/add-tag-dialog/add-tag-dialog";
 import { AboutComponent } from "../about-dialog/about-dialog";
+import { PieChartComponent } from "../pie-chart/pie-chart";
+import { BarChartComponent } from "../bar-chart/bar-chart";
+import { TableTransactionsComponent } from "../table-transactions/table-transactions";
+import { StorageService } from "../../services/storage";
```
Comment (gift-a) [inline src/components/app/app.js]: you need your own storage service

File: src/components/app/app.js [hunk 4]
```diff
@@ -35,23 +41,44 @@ export class AppComponent {
...
-    this.addTagDialog = new AddTagComponent(this.addTagMountPoint);
-    this.addTagDialog.mount();
     this.aboutDialog = new AboutComponent(this.aboutMountPoint);
     this.aboutDialog.mount();
   }
 
+  mountTransactionsComponent(tags) {
+    this.transactionsComponent = new TransactionsComponent(
+      this.transactionsMountPoint,
+      { tags: tags }
```
Comment (gift-a) [inline src/components/app/app.js]: what does this method do? You don't need to do anything with transaction component. your sphere of work is Drawer. TransactionsComponent is a content part. 
And by using this method you rewrite `this.transactionsComponent` from row 45

File: src/components/drawer/drawer.js [hunk 5]
```diff
@@ -15,38 +17,42 @@ export class DrawerComponent {
...
       ".drawer__add-account-dialog-activation"
     );
-    this.addTagButton = this.mountPoint.querySelector(
-      ".drawer__add-tag-dialog-activation"
-    );
-
-    this.moreBtnMountPoints = this.mountPoint.querySelectorAll(
-      ".drawer__more-button"
+    this.tagsMountPoint = this.mountPoint.querySelector(
+      ".drawer__tags-mountpoint"
```
Comment (gift-a) [inline src/components/drawer/drawer.js]: maybe `.drawer__tags-mount-point`?

File: src/components/app/app.js [hunk 6]
```diff
@@ -35,23 +38,35 @@ export class AppComponent {
...
     this.addAccountDialog = new AddAccountComponent(this.addAccountMountPoint);
     this.addAccountDialog.mount();
-    this.addTagDialog = new AddTagComponent(this.addTagMountPoint);
-    this.addTagDialog.mount();
     this.aboutDialog = new AboutComponent(this.aboutMountPoint);
     this.aboutDialog.mount();
   }
 
+  getTags() {
+    StorageService.get("tags").then(tags => {
```
Comment (ihor7016) [inline src/components/app/app.js]: you should not use storage service directly -- but make your wrapper around it to handle just tags

Comment (ihor7016) [reply src/components/app/app.js]: moreover -- why do you need this here? you are supposed to get your tags in `tags` component

File: src/model/tag.js [hunk 7]
```diff
@@ -0,0 +1,5 @@
+export class Tag {
```
Comment (ihor7016) [inline src/model/tag.js]: please remove it -- we do not use models

File: src/index.scss [hunk 8]
```diff
@@ -28,3 +28,4 @@
 @import "./components/button-more/button-more";
 @import "./components/table-transactions/table-transactions";
 @import "./components/chart/chart";
+@import "./components/tags/tags-component/tags-component";
```
Comment (ihor7016) [inline src/index.scss]: we do not use `component` in file name -- so it should be called just `tags`

File: src/components/drawer/drawer.js [hunk 9]
```diff
@@ -15,38 +17,42 @@ export class DrawerComponent {
...
   }
 
-  initMoreBtns() {
-    Array.from(this.moreBtnMountPoints).forEach(point => {
-      new ButtonMoreComponent(point, {
-        position: "right",
-        onDeleteClicked: this.handleDeleteClick.bind(this),
-        onEditClicked: this.handleEditClick.bind(this)
-      }).mount();
+  initTags(tags) {
```
Comment (ihor7016) [inline src/components/drawer/drawer.js]: please move this logic of working with tags to tags component

File: src/components/drawer/drawer.js [hunk 10]
```diff
@@ -15,38 +17,42 @@ export class DrawerComponent {
...
+  mountChildren() {
+    this.tagsComponent = new TagsComponent(this.tagsMountPoint, {
+      addTagMountPoint: this.props.addTagMountPoint,
+      onAddTagConfirmed: this.onAddTagConfirmed.bind(this)
+    });
+    this.tagsComponent.mount();
+  }
+
+  onAddTagConfirmed(tag) {
+    this.addTag(tag);
```
Comment (ihor7016) [inline src/components/drawer/drawer.js]: again these handlers for adding/removing tags should be in tags component

File: src/components/tags/tags-component/tags-component.js [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  initDialogComponent() {
+    this.addTagComponent = new AddTagComponent(this.props.addTagMountPoint, {
+      onAddTagConfirmed: this.handleAddTagConfirmed.bind(this)
+    });
+  }
+
+  handleAddTagConfirmed(tag) {
+    TagService.get().then(tags => {
```
Comment (ihor7016) [inline src/components/tags/tags-component/tags-component.js]: please move this logic to `TagService` to method called `add`

File: src/components/tags/tags-component/tags-component.js [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+      "click",
+      this.handleAddTagClick.bind(this)
+    );
+  }
+
+  addTag(tag) {
+    let element = document.createElement("div");
+    element.innerHTML = tagItemTemplate({ tag: tag });
+    let childNode = this.tagList.childNodes[2];
+    this.tagList.insertBefore(element.firstChild, childNode);
```
Comment (ihor7016) [inline src/components/tags/tags-component/tags-component.js]: better

```js
this.tagList.innerHTML = tagItemTemplate({ tag: tag }) + this.tagList.innerHTML
```

**Comment by ihor7016** [comment]
image

please store tags as just list of tag names like this
```js
['tag1', 'tag2', 'tag3']
```

File: src/components/app/app.js [hunk 13]
```diff
@@ -3,7 +3,8 @@ import template from "./app.html";
 import { DrawerComponent } from "../drawer/drawer";
 import { ToolbarComponent } from "../toolbar/toolbar";
 import { TransactionsComponent } from "../transactions/transactions";
-import { AddTagComponent } from "../add-tag-dialog/add-tag-dialog";
+import { AboutComponent } from "../about-dialog/about-dialog";
```
Comment (gift-a) [inline src/components/app/app.js:6]: it is in toolbar

File: src/components/app/app.js [hunk 14]
```diff
@@ -18,9 +19,10 @@ export class AppComponent {
     this.transactionsMountPoint = this.mountPoint.querySelector(
       ".app__container-content"
     );
-    this.addTagMountPoint = this.mountPoint.querySelector(
-      ".app__add-tag-dialog"
+    this.addAccountMountPoint = this.mountPoint.querySelector(
+      ".app__add-account-dialog"
     );
+    this.aboutMountPoint = this.mountPoint.querySelector(".app__about-dialog");
```
Comment (gift-a) [inline src/components/app/app.js:25]: it is in toolbar

File: src/components/app/app.js [hunk 15]
```diff
@@ -29,34 +31,37 @@ export class AppComponent {
...
+    this.addAccountDialog.mount();
+    
     this.transactionsComponent = new TransactionsComponent(
       this.transactionsMountPoint,
       {
         onTransactionAdded: this.handleTransactionAdded.bind(this)
       }
     );
     this.transactionsComponent.mount();
+    this.aboutDialog = new AboutComponent(this.aboutMountPoint);
```
Comment (gift-a) [inline src/components/app/app.js:44]: it is in toolbar

File: src/components/app/app.js [hunk 16]
```diff
@@ -29,34 +31,37 @@ export class AppComponent {
...
+    this.drawerComponent.toggleDrawer();
   }
 
   handleToolbarMenuClick() {
     this.drawerComponent.toggleDrawer();
   }
 
-  handleAddTagOnclick() {
-    this.addTagDialog.showDialog();
+  handleAboutOnclick() {
```
Comment (gift-a) [inline src/components/app/app.js:63]: it is in toolbar

File: src/components/drawer/drawer.js [hunk 17]
```diff
@@ -35,31 +38,15 @@ export class DrawerComponent {
...
-        position: "right",
-        onDeleteClicked: this.handleDeleteClick.bind(this),
-        onEditClicked: this.handleEditClick.bind(this)
-      }).mount();
-    });
-  }
-
   addEventListeners() {
-    this.addTagButton.addEventListener(
+    this.addAccountButton.addEventListener(
```
Comment (gift-a) [inline src/components/drawer/drawer.js:42]: why add account?

File: src/components/drawer/drawer.js [hunk 18]
```diff
@@ -35,31 +38,15 @@ export class DrawerComponent {
...
       "click",
-      this.handleAddTagOnclick.bind(this)
+      this.handleAddAccountClick.bind(this)
```
Comment (gift-a) [inline src/components/drawer/drawer.js:44]: why add account?

File: src/components/drawer/drawer.js [hunk 19]
```diff
@@ -35,31 +38,15 @@ export class DrawerComponent {
...
     );
   }
-  handleEditClick() {}
-
-  handleDeleteClick() {}
 
-  handleAddTagOnclick() {
-    this.props.onAddTagClick();
+  handleAddAccountClick() {
```
Comment (gift-a) [inline src/components/drawer/drawer.js:48]: why add account?

File: src/components/tags/tags.js [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+    Array.from(this.moreBtnMountPoints).forEach(point => {
+      new ButtonMoreComponent(point, {
+        position: "right",
+        onDeleteClick: this.handleDeleteClick.bind(this),
+        onEditClick: this.handleEditClick.bind(this)
+      }).mount();
+    });
+  }
+
+  initAddTagDialogComponent() {
```
Comment (gift-a) [inline src/components/tags/tags.js:47]: it should name `mountChildren`

**Comment by Vladyslav-Plakhuta** [comment]
removed 'model', made wrapper around storage service, moved logic of working with tags to 'tags component', renamed files in 'tags component'.

review pls

File: package.json [hunk 21]
```diff
@@ -52,6 +52,7 @@
     "@material/toolbar": "^0.32.0",
     "@material/typography": "^0.28.0",
     "chart.js": "^2.7.2",
-    "normalize.css": "^8.0.0"
+    "normalize.css": "^8.0.0",
+    "npm": "^5.7.1"
```
Comment (gift-a) [inline package.json:56]: why do you need `npm` here?

File: src/index.scss [hunk 22]
```diff
@@ -28,6 +28,4 @@
 @import "./components/button-more/button-more";
 @import "./components/table-transactions/table-transactions";
 @import "./components/chart/chart";
-@import "./components/accounts/accounts";
-@import "./components/custom-select/custom-select";
-@import "./components/about-dialog/about-dialog";
```
Comment (gift-a) [inline src/index.scss:33]: why do you delete this? you break app

File: src/components/drawer/drawer.js [hunk 23]
```diff
@@ -69,9 +56,8 @@ export class DrawerComponent {
   mount() {
     this.mountPoint.innerHTML = template();
     this.querySelectors();
-    this.initMoreBtns();
     this.initMDC();
     this.addEventListeners();
-    this.initAccountComponent();
```
Comment (gift-a) [inline src/components/drawer/drawer.js:75]: why do you delete this? it is not yours

File: src/components/app/app.js [hunk 24]
```diff
@@ -29,34 +31,37 @@ export class AppComponent {
...
 
   initDrawer() {
     this.drawerComponent = new DrawerComponent(this.drawerMountPoint, {
-      onAddTagClick: this.handleAddTagOnclick.bind(this)
+      onAddAccountClick: this.handleAddAccountClick.bind(this)
     });
     this.drawerComponent.mount();
   }
 
-  handleTransactionAdded(data) {
```
Comment (gift-a) [inline src/components/app/app.js:50]: why do you delete this? it is not yours

File: src/components/app/app.html [hunk 25]
```diff
@@ -5,6 +5,6 @@
 
     <div class="app__container-toolbar-point"></div>
     <div class="app__container-content"></div>
-    <div class="app__add-tag-dialog"></div>
+    <div class="app__add-account-dialog"></div>
```
Comment (gift-a) [inline src/components/app/app.html:8]: why account?

**Pull request closed without merging**